CONTRIBUTING.md

```diff
@@ -1,55 +1,123 @@
 
-### Branching model
+![clone.png](img/clone.png)
 
-- `master`: release branch, protected
-- Every feature/bugfix/... will have its own branch
-  - Branched off from `master`
-  - Pull request targeting the `master` branch
-  - Mandatory code review of the pull request
+- Complete the triangle workflow setup by adding `https://github.com/hivemq/mqtt-cli.git` as a remote named _upstream_ via IntelliJ by going to <i>Git > Manage Remotes > + </i> OR by using the command `git remote add upstream https://github.com/hivemq/mqtt-cli.git` 
 
-### Branching guidelines
+![manage-remote.png](img/manage-remote.png) 
 
-- Branch types: feature, bugfix, improvement, cleanup (same as the label of a corresponding GitHub Issue)
-- Branch names:
+
+## 2. Create a branch for your changes
+
+- Decide on your branch name:
+  - Branch types: feature, bugfix, improvement, cleanup (same as the label of a corresponding GitHub Issue)
   - Starting with type: `feature/`, `bugfix/`, `improvement/`, `cleanup/`
   - \+ task: lower case, spaces replaced with `-`
+- Create your branch by using IntelliJ's branch functionality at the bottom of the IDE (<i>upstream/master > New Branch from 'upstream/master'</i>) OR by using the command `git checkout -b branch-type/your-branch-name`
+
+![create-branch.png](img/create-branch.png)
+
+## 3. Write your code
+
+- Implement your changes under `src/main/java`
+- You can use the existing code as a guide
+- The project uses Nullability annotations to avoid NullPointerExceptions: `@NotNull`, `@Nullable`. 
+  Every non-primitive parameter/return type/field should be annotated with one of them.
+
+## 4. Write tests
+
+- There are three types of tests in mqtt-cli:
+  1. Unit Tests => `src/main/test`
+  2. Integration Tests => `src/main/integrationTest`
+  3. System Tests => `src/main/systemTest`
+- Your code changes must be unit-tested
+- Depending on your changes an integration test should be written
+- Depending on your changes or if you are developing a new feature a system test should be written
+
+## 5. Build and test
+
+- Before you build mqtt-cli, make sure that 
+  - a Docker daemon is running
+  - you have installed the native image tooling by using `./gradlew installNativeImageTooling`
+- To build mqtt-cli use `./gradlew build`
+- Before proceeding to the next step, please make sure that your changes build and all tests succeed
+- Tip: After building mqtt-cli you can execute the cli by using the generated jar file `java -jar build/libs/mqtt-cli-<version>.jar` (you can also just build this jar by using `./gradlew shadowJar`)
+- Tip: To quickly test your changes manually you can directly execute commands from your IDE by going to <i>Run > Edit Configurations...</i> and adding your cli arguments
+
+![run-config.png](img/run-config.png)
+
+
+## 6. Commit and Push
+- Before committing execute `./gradlew licenseFormat` to add license headers to each file
```

Comment:
Something like:
- Format each file you changed using the .editorconfig settings of the project.